airflow/contrib/hooks/ssh_hook.py

```
@@ -14,125 +14,141 @@
-
-        :param cmd: command to remotely execute
-        :return: stdout
-        """
-        p = self.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-        output, stderr = p.communicate()
-
-        if p.returncode != 0:
-            # I like this better: RemoteCalledProcessError(p.returncode, cmd, self.host, output=output)
-            raise AirflowException("Cannot execute {} on {}. Error code is: {}. Output: {}, Stderr: {}".format(
-                                   cmd, self.conn.host, p.returncode, output, stderr))
-
-        return output
+    def get_conn(self):
+        if not self.client:
+            logging.debug('creating ssh client for conn_id: {0}'.format(self.ssh_conn_id))
+            if self.ssh_conn_id is not None:
+                conn = self.get_connection(self.ssh_conn_id)
+                if self.username is None:
+                    self.username = conn.login
+                if self.password is None:
+                    self.password = conn.password
+                if self.remote_host is None:
+                    self.remote_host = conn.host
+                if conn.extra is not None:
+                    extra_options = conn.extra_dejson
+                    self.key_file = extra_options.get("key_file")
+
+                    if "timeout" in extra_options:
+                        self.timeout = int(extra_options["timeout"], 10)
+
+                    if "compress" in extra_options \
+                            and extra_options["compress"].lower() == 'false':
+                        self.compress = False
+                    if "no_host_key_check" in extra_options \
+                            and extra_options["no_host_key_check"].lower() == 'false':
+                        self.compress = False
+
+            if not self.remote_host:
+                raise AirflowException("Missing required param: remote_host")
+
+            # Auto detecting some values from system
+            if not self.username:
```

@jhsenjaliya I think it would be better to raise an exception instead, as the username to get with getpass is not necessary the correct one to use.

---

so I added this based on the thought that this is actually the last try to get username after trying to get it from Connection and params, and will fail if its not intended to, but gives a little convenience so that username doesn't has to be provided in cases where the airflow process is running as intended user.

---

anyway, I think it depends on use-case and may not work as generalized solution for all, so i ll take your suggestion and remove it.  Thanks !